src/QMCHamiltonians/MomentumEstimator.cpp

```diff
@@ -172,115 +147,271 @@ void MomentumEstimator::setParticlePropertyList(PropertySetType& plist
 {
   OhmmsAttributeSet pAttrib;
   std::string hdf5_flag="yes";
+  kmax=0.0;
+  kmax0=0.0;
+  kmax1=0.0;
+  kmax2=0.0;
   pAttrib.add(hdf5_flag,"hdf5");
-  pAttrib.add(kgrid,"grid");
-  pAttrib.add(M,"samples");
+  //pAttrib.add(kgrid,"grid");
+  pAttrib.add(kmax,"kmax");
+  pAttrib.add(kmax0,"kmax0");
+  pAttrib.add(kmax1,"kmax1");
+  pAttrib.add(kmax2,"kmax2");
+  pAttrib.add(M,"samples"); // default value is 40 (in the constructor)
   pAttrib.put(cur);
   hdf5_out = (hdf5_flag=="yes");
+  RealType min_Length=1.0e10;
+  PosType vec_length(0);
+  for (int i=0; i<OHMMS_DIM; i++)
+  {
+    PosType a(0);
+    for (int j=0; j<OHMMS_DIM; j++)
+      a[j]=elns.Lattice.G(j,i);
+    //length of reciprocal lattice vector
+    vec_length[i]=2.0*M_PI*std::sqrt(dot(a,a));
+    if (min_Length>vec_length[i])
+      min_Length=vec_length[i];
+  }
 #if OHMMS_DIM==3
-  int numqtwists(6*kgrid+3);
-  std::vector<int> qk(0);
-  mappedQtonofK.resize(numqtwists,qk);
-  compQ.resize(numqtwists);
-  RealType qn(4.0*M_PI*M_PI*std::pow(Lattice.Volume,-2.0/3.0));
-  mappedQnorms.resize(numqtwists,qn*0.5/RealType(M));
-  if (twist[0]==0)
-    mappedQnorms[kgrid]=qn/RealType(M);
-  if (twist[1]==0)
-    mappedQnorms[3*kgrid+1]=qn/RealType(M);
-  if (twist[2]==0)
-    mappedQnorms[5*kgrid+2]=qn/RealType(M);
-  Q.resize(numqtwists);
-  for (int i=-kgrid; i<(kgrid+1); i++)
+  PosType kmaxs(0);
+  kmaxs[0]=kmax0;
+  kmaxs[1]=kmax1;
+  kmaxs[2]=kmax2;
+  bool sphere_only=true;
+  bool directional_only=true;
+  RealType sum_kmaxs=kmaxs[0]+kmaxs[1]+kmaxs[2];
+  RealType sphere_kmax;
+  if (kmax<1.0e-10 && sum_kmaxs<1.0e-10)
   {
-    PosType kpt;
-    kpt[0]=i-twist[0];
-    kpt[1]=i-twist[1];
-    kpt[2]=i-twist[2];
-    kpt=Lattice.k_cart(kpt);
-    Q[i+kgrid]=std::abs(kpt[0]);
-    Q[i+kgrid+(2*kgrid+1)]=std::abs(kpt[1]);
-    Q[i+kgrid+(4*kgrid+2)]=std::abs(kpt[2]);
+    // default: kmax = 2 x k_F of polarized non-interacting electron system
+    kmax = 2.0*std::pow(6.0*M_PI*M_PI*elns.getTotalNum()/elns.Lattice.Volume,1.0/3);
+    kgrid = int(kmax/min_Length)+1;
+    sphere_kmax=kmax;
+    sphere_only=true;
+    directional_only=false;
+  }
+  else
+  {
+    sphere_kmax=kmax;
+    if (sum_kmaxs>0.0 && kmax>0.0)
+    {
+      directional_only=false;
+      sphere_only=false;
+    }
+    else
+    {
+      if (kmax<1.0e-10)
+        sphere_only=false;
+      if (sum_kmaxs<1.0e-10)
+        directional_only=false;
+    }
+    for (int i=0; i<OHMMS_DIM; i++)
+    {
+      if (kmaxs[i]>kmax)
+        kmax=kmaxs[i];
+    }
+    kgrid = int(kmax/min_Length)+1;
   }
-  app_log()<<" Using all k-space points with (nx^2+ny^2+nz^2)^0.5 < "<< kgrid <<" for Momentum Distribution."<< std::endl;
-  app_log()<<"  My twist is:"<<twist[0]<<"  "<<twist[1]<<"  "<<twist[2]<< std::endl;
-  int indx(0);
-  int kgrid_squared=kgrid*kgrid;
+  int kgrid_squared[OHMMS_DIM];
+  for (int i=0; i<OHMMS_DIM; i++)
+    kgrid_squared[i]=int(std::round(kmaxs[i]*kmaxs[i]/vec_length[i]/vec_length[i]));
```

Comment:
should be float and no int(round())

Reply:
I agree that the kgrid* variables could be better named as ngrid*, but I followed the old convention in which these are actually integers.

Reply:
I was thinking of twists. Since the twists will be added later. rounding here will cause bugs.